torchdata/stateful_dataloader/incremental_state.py

```diff
@@ -0,0 +1,178 @@
+                    continue
+            except Exception:
+                # Fallback to retaining new key/value
+                pass
+            delta_flat_state[key] = new_value
+        # Update internal state to the new state
+        self.flat_state = new_flat_state
+        return delta_flat_state
+
+    def apply_delta(self, flat_delta_state: Dict[str, Any]) -> None:
+        for key, update in flat_delta_state.items():
+            if self.flat_state is None:
+                self.flat_state = {}
+
+            if isinstance(update, DeletionTombStone):
+                # Remove key if present in the state
+                self.flat_state.pop(key, None)
+            else:
+                self.flat_state[key] = update
+
+    def get_state(self) -> Optional[Dict[str, Any]]:
+        return Flattener.unflatten(self.flat_state)
+
+
+class IncrementalWorkerState:
+    def __init__(self, initial_worker_state_dict: Optional[Dict[str, Any]]):
+        self._worker_id = None
+        self._fetcher_ended = None
+
+        dataset_state = None
+        fetcher_iter_state = None
+        if initial_worker_state_dict:
+            self._worker_id = initial_worker_state_dict[_WORKER_ID]
+            dataset_state = initial_worker_state_dict.get(_DATASET_STATE, None)
+            if fetcher_state := initial_worker_state_dict.get(_FETCHER_STATE, None):
+                self._fetcher_ended = fetcher_state[_FETCHER_ENDED]
+                fetcher_iter_state = fetcher_state.get(_DATASET_ITER_STATE, None)
+
+        self._incr_dataset_state = IncrementalState(dataset_state)
+        self._incr_fetcher_iter_state = IncrementalState(fetcher_iter_state)
+
+    def generate_delta(self, new_state_dict: Dict[str, Any]) -> Dict[str, Any]:
+        assert _WORKER_ID in new_state_dict
+        self._worker_id = new_state_dict[_WORKER_ID]
+        incr_state_dict = {_WORKER_ID: self._worker_id, _FETCHER_STATE: None}
+
+        if ds_state := new_state_dict.get(_DATASET_STATE, None):
+            incr_state_dict[_DATASET_STATE] = self._incr_dataset_state.generate_delta(ds_state)
+
+        if fetcher_state := new_state_dict.get(_FETCHER_STATE, None):
+            self._fetcher_ended = fetcher_state[_FETCHER_ENDED]
+
+            delta_iter_state = None
+            if iter_state := fetcher_state.get(_DATASET_ITER_STATE, None):
+                delta_iter_state = self._incr_fetcher_iter_state.generate_delta(iter_state)
+
+            incr_state_dict[_FETCHER_STATE] = {
+                _DATASET_ITER_STATE: delta_iter_state,
+                _FETCHER_ENDED: self._fetcher_ended,
+            }
+        return incr_state_dict
+
+    def apply_delta(self, delta_state_dict: Dict[str, Any]) -> None:
+        self._worker_id = delta_state_dict[_WORKER_ID]
+        if ds_state := delta_state_dict.get(_DATASET_STATE, None):
+            self._incr_dataset_state.apply_delta(ds_state)
+
+        if fetcher_state := delta_state_dict.get(_FETCHER_STATE, None):
```

Comment:
note: walrus requires Python >= 3.8, not sure if we are supporting older python versions

Reply:
pytorch core recommendation seems to be 3.8+ - https://pytorch.org/get-started/locally/#linux-python. Also our CI runs only for 3.8-3.10. So I think this is alright?

Reply:
makes sense, just wanted to flag so we know explicitly we have 3.8+ only features